src/wp-includes/rest-api/endpoints/class-wp-rest-application-passwords-controller.php

```diff
@@ -707,7 +707,7 @@ protected function get_user( $request ) {
 			return $error;
 		}
 
-		if ( is_multisite() && ! is_user_member_of_blog( $user->ID ) ) {
+		if ( is_multisite() && ! user_can( $user->ID, 'manage_sites' ) && ! is_user_member_of_blog( $user->ID ) ) {
```

Comment:
This code was inspired by code already in `WP_REST_Users_Controller`:

https://github.com/johnbillion/wordpress-develop/blob/953446d4d6e388b75b0f22005c9643c9811a3aca/src/wp-includes/rest-api/endpoints/class-wp-rest-users-controller.php#L391-L399

In general, `is_user_member_of_blog()` does not work well as a capability check, and probably should be swapped out for one nearly everywhere. (Conceptually, a user could be a member of a site with a role full of falsey capabilities to explicitly block a user from a site – they exist, but without the intended access. bbPress does this, for example.)

Internally, `wp_admin_bar_site_menu()` uses `current_user_can( 'manage_network' )` which maybe matches better than `manage_sites`?

Tangentially, the method being changed here starts with `wp_is_application_passwords_available()` (to put out its own error) but it gets called internally a second time by `wp_is_application_passwords_available_for_user()` so there is opportunity for future optimization here.

Reply:
> In general, is_user_member_of_blog() does not work well as a capability check, and probably should be swapped out for one nearly everywhere. (Conceptually, a user could be a member of a site with a role full of falsey capabilities to explicitly block a user from a site – they exist, but without the intended access. bbPress does this, for example.)

I don't think `is_user_member_of_blog` is being used as a capability check. We're just trying to restrict the user to being accessed using the REST API on a site they are a member of. But App Passwords are global user data, so I'm not sure what site-specific capability would make sense here.

> Tangentially, the method being changed here starts with wp_is_application_passwords_available() (to put out its own error) but it gets called internally a second time by wp_is_application_passwords_available_for_user() so there is opportunity for future optimization here.

This is intentional so we can provide better error messages.

> Tangentially again, get_user_by() and get_userdata() have their own internal check for the $current_user global, so the me check could use get_current_user_id() instead of wp_get_current_user() unless it needs everything inside of _wp_get_current_user(). Wee!

I'm not sure what you are proposing here? `get_current_user_id()` calls `wp_get_current_user`.

Reply:
> I don't think `is_user_member_of_blog` is being used as a capability check. We're just trying to restrict the user to being accessed using the REST API on a site they are a member of.

...which is a capability check. 😬 They could have a `blocked` bbPress role, making them "a member of that site" that still shouldn't be able to manage application passwords IMO.

> But App Passwords are global user data, so I'm not sure what site-specific capability would make sense here.

Why restrict access to users who are members of a site, and why send back "Invalid user ID." when the user ID _is_ valid but their lack of a role on that site is kicking the request back?

Shouldn't a user that is global in the system be able to hit `/me` on _any_ site and get something back?

> This is intentional so we can provide better error messages.

I understand. It is still, however, redundant.

> I'm not sure what you are proposing here?

From this:
```php
		if ( 'me' === $id ) {
			if ( ! is_user_logged_in() ) {
				return new WP_Error(
					'rest_not_logged_in',
					__( 'You are not currently logged in.' ),
					array( 'status' => 401 )
				);
			}

			$user = wp_get_current_user();
		} else {
			$id = (int) $id;

			if ( $id <= 0 ) {
				return $error;
			}

			$user = get_userdata( $id );
		}
```

To this:
```php
		if ( 'me' === $id ) {
			if ( ! is_user_logged_in() ) {
				return new WP_Error(
					'rest_not_logged_in',
					__( 'You are not currently logged in.' ),
					array( 'status' => 401 )
				);
			}

			$id = get_current_user_id();
		} else {
			$id = (int) $id;

			if ( $id <= 0 ) {
				return $error;
			}
		}

		$user = get_userdata( $id );
```

There's 2 steps above:
* Resolve the user ID for the request ("me" vs. int)
* Retrieve user data for that ID

Normally, I'd suggest that the code do its best to communicate the separation of those logical steps.

Is the `me` check appears to be unique to this endpoint? I think so, and I see that `wp-api.js` has some special consideration for it too. It's a weird unique code juggle.

Reply:
> ...which is a capability check. 😬 They could have a blocked bbPress role, making them "a member of that site" that still shouldn't be able to manage application passwords IMO.

This entire code block isn't a permissions check though. This function is just to identify the user being requested. Each individual action has its own specific capabilities check that bbPress or any other plugin could modify behavior for.

> Shouldn't a user that is global in the system be able to hit /me on any site and get something back?

It could work that way, but it was considered a bug in [#39701](https://core.trac.wordpress.org/ticket/39701).

---

I guess I'll just say I disagree here that changing the code block from producing a user object to a user id is clearer.

Reply:
I'd like to go ahead with my proposed change here and recommend that if any changes need to be made to the use of `is_user_member_of_blog()` here then it be done in a follow up ticket. Otherwise the fix for this rather annoying bug is going to stagnate.

Any objections to committing this change?

Reply:
Current patch looks good to me.